nemo/collections/tts/modules/submodules.py

```diff
@@ -507,3 +514,240 @@ def forward(self, inputs, conditioning=None):
+        )
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inputs_lengths": NeuralType(('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "out": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs, inputs_lengths):
+        # BMW -> BWMC (M: mels)
+        x = inputs.transpose(1, 2).unsqueeze(3)
+        x_lens = inputs_lengths
+        x_masks = self.lengths_to_masks(x_lens).unsqueeze(2).unsqueeze(3)
+
+        for layer in self.layers:
+            x = layer(x, x_masks)
+            x_lens = self.calculate_post_conv_lengths(x_lens)
+            x_masks = self.lengths_to_masks(x_lens).unsqueeze(2).unsqueeze(3)
+
+        # BWMC -> BWC
+        x = x.contiguous().view(x.shape[0], x.shape[1], -1)
+
+        self.gru.flatten_parameters()
+        packed_x = pack_padded_sequence(x, x_lens.cpu(), batch_first=True, enforce_sorted=False)
+        packed_x, _ = self.gru(packed_x)
+        x, x_lens = pad_packed_sequence(packed_x, batch_first=True)
+        x = x[torch.arange(len(x_lens)), (x_lens - 1), :]
+        return x
+
+    @staticmethod
+    def calculate_post_conv_lengths(lengths, n_convs=1, kernel_size=3, stride=2, pad=1):
+        """Batch lengths after n convolution with fixed kernel/stride/pad."""
+        for _ in range(n_convs):
+            lengths = (lengths - kernel_size + 2 * pad) // stride + 1
+        return lengths
+
+    @staticmethod
+    def lengths_to_masks(lengths):
+        """Batch of lengths to batch of masks"""
+        # B -> BxT
+        masks = torch.arange(lengths.max()).to(lengths.device).expand(
+            lengths.shape[0], lengths.max()
+        ) < lengths.unsqueeze(1)
+        return masks
+
+
+class GlobalStyleToken(NeuralModule):
+    """
+    Global Style Token based Speaker Embedding
+    """
+
+    def __init__(
+        self, reference_encoder, gst_size=128, n_style_token=10, n_style_attn_head=4,
+    ):
+        super(GlobalStyleToken, self).__init__()
+        self.reference_encoder = reference_encoder
+        self.style_attention = StyleAttention(
+            gst_size=gst_size, n_style_token=n_style_token, n_style_attn_head=n_style_attn_head
+        )
+
+    @property
+    def input_types(self):
+        return {
+            "inp": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType()),
+            "inp_lengths": NeuralType(('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "gst": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inp, inp_lengths):
+        style_embedding = self.reference_encoder(inp, inp_lengths)
+        gst = self.style_attention(style_embedding)
+        return gst
+
+
+class LookupTable(torch.nn.Module):
+    """
+    LookupTable based Speaker Embedding
+    """
```

Comment:
Maybe rename it to SpeakerLookupTable to make it clear how it differs from torch.nn.Embedding.